src/dasi/api/dasi_c.h

```diff
@@ -0,0 +1,235 @@
+typedef struct Query dasi_query_t;
+
+struct dasi_list_t;
+typedef struct dasi_list_t dasi_list_t;
+
+struct dasi_retrieve_t;
+typedef struct dasi_retrieve_t dasi_retrieve_t;
+
+/* ---------------------------------------------------------------------------------------------------------------------
+ * ERROR HANDLING
+ * -------------- */
+
+/* DASI Error Codes */
+
+typedef enum dasi_error_values_t
+{
+    DASI_SUCCESS            = 0, /* Operation succeded. */
+    DASI_ITERATION_COMPLETE = 1, /* All elements have been returned */
+    DASI_ERROR              = 2, /* Operation failed. */
+    DASI_ERROR_UNKNOWN      = 3, /* Failed with an unknown error. */
+    DASI_ERROR_USER         = 4, /* Failed with an user error. */
+    DASI_ERROR_ITERATOR     = 5, /* Failed with an iterator error. */
+    DASI_ERROR_ASSERT       = 6  /* Failed with an assert() */
+} dasi_error_enum_t;
+
+const char* dasi_get_error_string(int err);
+
+/* -----------------------------------------------------------------------------
+ * HELPERS
+ * ------- */
+
+/**
+ * @brief Set DASI version.
+ *
+ * @param version Version string
+ * @return int Error code
+ */
+int dasi_version(const char** version);
+
+/**
+ * @brief Set DASI git sha1 version.
+ *
+ * @param sha1 SHA1 version string
+ * @return int Error code
+ */
+int dasi_vcs_version(const char** sha1);
+
+/**
+ * @brief Initialise Main() context.
+ *
+ * @note This is ONLY required when Main() is NOT initialised, such as loading
+ * the DASI as shared library in Python.
+ */
+int dasi_initialise_api();
+
+/* ---------------------------------------------------------------------------------------------------------------------
+ * DASI SESSION
+ * ------------ */
+
+/** Create a new session object using the given configuration file. */
+int dasi_open(dasi_t** dasi, const char* filename);
+
+/** Release the session and delete the object. */
+int dasi_close(const dasi_t* dasi);
+
+/**
+ * @brief Writes data to the object store.
+ *
+ * @note Data is not guaranteed accessible nor persisted (wrt. failure),
+ * until dasi_flush() is called.
+ *
+ * @param dasi Dasi session
+ * @param key Metadata description of the data to store and index
+ * @param data Pointer to the read-only data
+ * @param length Length of "data" in bytes
+ */
+int dasi_archive(dasi_t* dasi, const dasi_key_t* key, const void* data, long length);
+
+int dasi_flush(dasi_t* dasi);
+
+/* *** List functionality */
+
+int dasi_new_list(dasi_t* dasi, const dasi_query_t* query, dasi_list_t** list);
+
+int dasi_free_list(const dasi_list_t* list);
+
+int dasi_list_next(dasi_list_t* list);
+
+int dasi_list_attrs(const dasi_list_t* list,
+                    dasi_key_t** key,
+                    dasi_time_t* timestamp,
+                    const char** uri,
+                    long* offset,
+                    long* length);
+
+/* *** Retrieve functionality */
+
+int dasi_new_retrieve(dasi_t* dasi, const dasi_query_t* query, dasi_retrieve_t** retrieve);
```

Comment:
Same comment here.

dasi_new_retrieve --> dasi_retrieve (which is the function you are doing).

Reply:
same reason with list; this allocate memory and naming convention is to add new as indicator.